src/machine.rs

```diff
@@ -404,6 +412,60 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
         Ok(())
     }
 
+    fn eval_maybe_thread_local_static_const(
+        ecx: &InterpCx<'mir, 'tcx, Self>,
+        mut val: mir::interpret::ConstValue<'tcx>,
+    ) -> InterpResult<'tcx, mir::interpret::ConstValue<'tcx>> {
+        match &mut val {
+            mir::interpret::ConstValue::Scalar(Scalar::Ptr(ptr)) => {
+                let alloc_id = ptr.alloc_id;
+                let alloc = ecx.tcx.alloc_map.lock().get(alloc_id);
+                let tcx = ecx.tcx;
+                let is_thread_local = |def_id| {
+                    tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::THREAD_LOCAL)
+                };
+                match alloc {
+                    Some(GlobalAlloc::Static(def_id)) if is_thread_local(def_id) => {
+                        let new_alloc_id = if let Some(new_alloc_id) =
+                            ecx.get_thread_local_alloc_id(alloc_id)
+                        {
+                            new_alloc_id
+                        } else {
+                            if tcx.is_foreign_item(def_id) {
+                                throw_unsup_format!(
+                                    "Foreign thread-local statics are not supported."
+                                )
+                            }
+                            let instance = Instance::mono(tcx.tcx, def_id);
+                            let gid = GlobalId { instance, promoted: None };
+                            let raw_const = tcx
+                                .const_eval_raw(ty::ParamEnv::reveal_all().and(gid))
```

Comment:
Ouch, having to access the query directly is bad. Also the code here looks slightly different from `InterpCtx::const_eval` in the core engine, why is that?

If we cannot somehow share this code with the core engine, please add a FIXME to revisit that once https://github.com/rust-lang/rust/issues/70685 is resolved.

Reply:
(In other words, this is another layering violation -- the Miri tool never should have to dig down so deep into rustc.)

Reply:
> Also the code here looks slightly different from InterpCtx::const_eval in the core engine, why is that?

This code is copied from `Memory::get_global_alloc`.